### PR TITLE
added extractScientificMetadata and tools

### PR DESCRIPTION
**Pull request by toqduj**
Added tools for dealing with nexus and generic HDF5 files. 
In particular, adds extractScientificMetadata method that traverses an HDF5 tree and returns a dict with all the metadata and their associated units. This dict can be directly used to fill out the scientific metadata tag in a SciCat dataset. 

Also convenient could be the H5Get and H5GetDict methods that can be used to easily extract values from an HDF5 file with error and default handling.

### CODE REVIEW

File: requirements-hdf5.txt [hunk 1]
```diff
@@ -0,0 +1,5 @@
+pydantic
+requests
```
Comment (dylanmcreynolds) [inline requirements-hdf5.txt:2]: A few notes about this...
`pydantic` and `requests` are already being installed through requirements, so they are not needed here.

Also, I don't think you need to import `logging` into python any more, unless you have to support very  old versions of python.

**Comment by dylanmcreynolds** [comment]
Thanks for the PR!  This is exciting.

This repo uses a pre-commit step in github to check formatting through flake8 and black. Currently the PR is not passing both. Can you take a look? https://github.com/SciCatProject/pyscicat/runs/4797593179?check_suite_focus=true

Also, I would really like to see some unit tests (with pytest) before merging it into pyscicat. Perhaps add a folder called `_tests` in ` hdf5scicattools`?  (In `pyscicat` I called it `tests` but will probably change that soon.)

It would also be great if you could add to the user manual under `docs/source/howto`.  

Thanks!

**Comment by toqduj** [comment]
Hi Dylan, 

I've added some tests (did one in unittest, as I'm more familiar with that one, would that be a problem?) and a very small test dataset. However, I'll still need to work on the docs and (Black) formatting.

**Comment by toqduj** [comment]
Alright, should now be passing black and flake8. The latter gave an importwarning for hdf5plugin (F401) which I've suppressed. It appears to be an unused library but actually extends the previously loaded h5py library with additional compression algorithms. docs are coming.

**Comment by toqduj** [comment]
some documentation there, but I haven't added an example in the documentation yet. Please let me know if you agree with the documentation location etc.

**Comment by dylanmcreynolds** [comment]
> I can change that line to `from .h5tools import h5py_casting`. After that, jupyter notebook works fine, but testing in Visual Studio Code then complains and wants me to remove that relative import . again. Any ideas?

`from .h5tools import h4_casting` seems appropriate to me. h5tools is a relative import. Two related suggestions:

1.  My experience with VSCode is that it can be slow to picking up package changes from the interpreter environment. I find myself sometimes restarting it or changing the python interpreter to something  bogus and then back (using SHFT+CTL+P and "Python: Select Interpreter" or clicking on the python interpreter listing in the bottom of the screen, which does the same)
2. If you haven't done so in your environment, you'll want to install the pyscicat source into your intrepter. From the root of pyscicat, type `pip install -e .`. This will install the packages. 

So, if you haven't done 2, try that and then try 1.

**Comment by dylanmcreynolds** [comment]
> So how does this work, do I need to make a new PR or kick off a new review somehow?
Nope. We can keep this PR going until it's merged.

File: hdf5scicattools/_tests/hdf5sct_unittest.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
+import unittest
+
+# these need to be loaded at the beginning to avoid errors related to relative imports (ImportWarning in h5py)
+# might be related to the change of import style for Python 3.5+. Tested on Python 3.7 at 20200417
```
Comment (dylanmcreynolds) [inline hdf5scicattools/_tests/hdf5sct_unittest.py:4]: Currently, pyscicat only supports 3.7, 3.8, 3.9.

**Comment by toqduj** [comment]
Documentation added, and unittest changed to reflect the relative imports

File: hdf5scicattools/_tests/hdf5sct_unittest.py (hunk 2 above)
Comment (toqduj) [reply hdf5scicattools/_tests/hdf5sct_unittest.py]: whoops, leftover from old donor code.

**Comment by dylanmcreynolds** [comment]
> some documentation there, but I haven't added an example in the documentation yet. Please let me know if you agree with the documentation location etc.

This is fine for now. Sadly, `ingesting.md` that is an older file that I forgot to incorporate in the official  sphynx-built documentation, so it's currently there but not built into anything. I'll address that. I would still love to see something in the how-to, if possible. But we could do that in a separate PR.

**Comment by dylanmcreynolds** [comment]
It is theoretically possible to have `pytest` run `unittest`. But I would really prefer to not incur the extra hassle in CI. 

Below is a quick conversion of your tests from `unittest` to `pytest`. To get pytest to recognize the test, the filename has to have `test_` prepended.

Then to get that to run from a command line, just type `pytest`. To get that to run in the VSCode test tools, you may need to tell vscode to use pytest.

```
from pathlib import Path

from ..extractscientificmetadata import extractScientificMetadata
from ..h5tools import h5Get, h5GetDict

def test_readValue(self):
    p = Path(
        "hdf5scicattools/_tests/testdata/cylinderHex_r5_s12_T50_large_ranW_0p5.nxs"
    )
    v = h5Get(p, "/sasentry1/sasdata1/I")
    self.assertTrue(v != "none", "Did not extract value")

def test_readAttribute(self):
    p = Path(
        "hdf5scicattools/_tests/testdata/cylinderHex_r5_s12_T50_large_ranW_0p5.nxs"
    )
    v = h5Get(p, "/sasentry1/sasdata1@timestamp")
    self.assertTrue(v != "none", "Did not extract attribute")

def test_readMixedDict(self):
    p = Path(
        "hdf5scicattools/_tests/testdata/cylinderHex_r5_s12_T50_large_ranW_0p5.nxs"
    )
    v = h5GetDict(
        p,
        {
            "/sasentry1/sasdata1@timestamp": 123,
            "/sasentry1/sasdata1/I": 0,
            "/sasentry1/sasdata1/I@units": "inverse amps",
        },
    )
    self.assertTrue(
        v["/sasentry1/sasdata1/I@units"] != "inverse amps",
        "Did not extract unit attribute",
    )
    self.assertTrue(
        v["/sasentry1/sasdata1/I"] != 0, "Did not extract intensity value(s)"
    )

def test_readMetadata_withroot(self):
    p = Path(
        "hdf5scicattools/_tests/testdata/cylinderHex_r5_s12_T50_large_ranW_0p5.nxs"
    )
    assert p.exists(), f"HDF5/NeXus test file: {p.as_posix()} cannot be found"
    resultDict = extractScientificMetadata(
        p, excludeRootEntry=True, skipKeyList=["sasdata1"]
    )
    self.assertIsNotNone(
        resultDict, "extractScientificMetadata has not returned anything"
    )

    # make sure the root entry has been skipped
    self.assertTrue(
        "sasentry1" not in resultDict.keys(), "Root entry was not excluded"
    )

    # make sure the skipKeyList item has been skipped
    self.assertTrue("sasdata1" not in resultDict.keys(), "skipKey was not skipped")
```

**Comment by toqduj** [comment]
I'll modify the tests. 

Just not 100% sure about the correct location for this package. Is it better here, alongside the pyscicat subdirectory, or should it be in the pyscicat subdirectory?

**Comment by toqduj** [comment]
Ok, tests changed and passing.

**Comment by dylanmcreynolds** [comment]
I hate to be passing code in the comments, but I don't have permission to commit to your branch. Here's what I'd like to see in setup.py, which lets you install either:
`pip install pyscicat[base]`
`pip install pyscicat[h5tools]`

```
from os import path
from setuptools import setup, find_packages
import sys
import versioneer

min_version = (3, 7)
if sys.version_info < min_version:
    error = """
pyscicat does not support Python {0}.{1}.
Python {2}.{3} and above is required. Check your Python version like so:

python3 --version

This may be due to an out-of-date pip. Make sure you have pip >= 9.0.1.
Upgrade pip like so:

pip install --upgrade pip
""".format(
        *(sys.version_info[:2] + min_version)
    )
    sys.exit(error)

here = path.abspath(path.dirname(__file__))
with open(path.join(here, "README.md"), encoding="utf-8") as readme_file:
    readme = readme_file.read()

with open(path.join(here, "requirements.txt")) as requirements_file:
    # Parse requirements.txt, ignoring any commented-out lines.
    requirements = [
        line
        for line in requirements_file.read().splitlines()
        if not line.startswith("#")
    ]

with open(path.join(here, "requirements-hdf5.txt")) as requirements_hdf5_file:
    # Parse requirements.txt, ignoring any commented-out lines.
    requirements_hdf5 = [
        line
        for line in requirements_hdf5_file.read().splitlines()
        if not line.startswith("#")
    ]

extras_require = {}
extras_require["base"] = requirements
extras_require["h5tools"] = requirements_hdf5

setup(
    name="pyscicat",
    version=versioneer.get_version(),
    cmdclass=versioneer.get_cmdclass(),
    description="Code for communicating to a SciCat backend server python",
    long_description=readme,
    author="Dylan McReynolds",
    author_email="dmcreynolds@lbl.gov",
    url="https://github.com/scicatproject/pyscicat",
    python_requires=">={}".format(".".join(str(n) for n in min_version)),
    packages=find_packages(exclude=["docs", "_tests"]),
    include_package_data=True,
    install_requires=[],
    extras_require=extras_require,
    license="BSD (3-clause)",
    classifiers=[
        "Development Status :: 2 - Pre-Alpha",
        "Natural Language :: English",
        "Programming Language :: Python :: 3",
    ],
)```

**Comment by dylanmcreynolds** [comment]
> I'll modify the tests.
> 
> Just not 100% sure about the correct location for this package. Is it better here, alongside the pyscicat subdirectory, or should it be in the pyscicat subdirectory?

Yeah, I go back and forth, too. I think I like under `pyscicat` better than as a sibling, though. It would be more standard. 
It would be more natural, I think, to `import pyscicat.h5tools` rather than `import hdfscicattools`

**Comment by toqduj** [comment]
yea, I think it's better under pyscicat rather than alongside too. I'll move it and see if I can get it to work with those imports..

File: pyscicat/hdf5/extractscientificmetadata.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+        if list(attributes.keys()) == ["value"]:  # no attributes here
+            nested_dict = val
+        else:
+            nested_dict = attributes.copy()
+        if val != "":
+            update_dict = build_dictionary(levels, nested_dict)
+            metadata = update_deep(metadata, update_dict)
+
+
+def extractScientificMetadata(
```
Comment (dylanmcreynolds) [inline pyscicat/hdf5/extractscientificmetadata.py:94]: To match PEP naming conventions, suggest extract_scientific_metdata for this function.

**Comment by dylanmcreynolds** [comment]
What would you think of renaming `extractcientificmetadata.py` to 'scientific_metadata.py`?

**Comment by dylanmcreynolds** [comment]
@toqduj, the tests are failing because the CI server is not installing your new requirements. To fix this, go here: https://github.com/SciCatProject/pyscicat/blob/9bc608eb14110bbd9302724f6fd1083359c66619/.github/workflows/testing.yml#L34

Add a line to install your requirements file, so it looks like:
```      run: |
        set -vxeuo pipefail
        python -m pip install -r requirements-dev.txt
        python -m pip install -r requirements-hdf5.txt
        python -m pip list```

**Comment by dylanmcreynolds** [comment]
I went ahead an merged. I'll take care of the pytest myself. Thanks for the contribution!

**Pull request merged**